core/aws-core/src/main/java/software/amazon/awssdk/awscore/rules/AwsEndpointProviderUtils.java

```diff
@@ -154,6 +160,91 @@ public static SdkHttpRequest setUri(SdkHttpRequest request, URI clientEndpoint,
                       .build();
     }
 
+    public static AwsRequest addHeaders(AwsRequest request, Map<String, List<String>> headers) {
+        AwsRequestOverrideConfiguration.Builder configBuilder = request.overrideConfiguration()
+                                                                       .map(AwsRequestOverrideConfiguration::toBuilder)
+                                                                       .orElseGet(AwsRequestOverrideConfiguration::builder);
+
+
+        headers.forEach((name, values) -> {
+            List<String> existingValues = configBuilder.headers().get(name);
+            List<String> updatedValues;
+
+            if (existingValues != null) {
+                updatedValues = new ArrayList<>(existingValues);
+            } else {
+                updatedValues = new ArrayList<>();
+            }
+
+            updatedValues.addAll(values);
+
+            configBuilder.putHeader(name, updatedValues);
+        });
+
+        return request.toBuilder()
+            .overrideConfiguration(configBuilder.build())
+            .build();
+    }
+
+    /**
+     * Per the spec, the auth schemes list is ordered by preference, so we simply iterate over the list until we find an
+     * auth scheme we recognize.
+     */
+    public static EndpointAuthScheme chooseAuthScheme(List<EndpointAuthScheme> authSchemes) {
+        for (EndpointAuthScheme authScheme : authSchemes) {
+            if (SIGV4_NAME.equals(authScheme.name()) || SIGV4A_NAME.equals(authScheme.name())) {
```

Comment:
I'm not sure I see the need for adding the named variables like `SIGV4_NAME`. If we get a third schema then we need to add more code. Can we take the list of supported auth schemes and compare with the resolved auth schemes and pick the right one?